singer-commons/src/main/java/com/pinterest/singer/loggingaudit/client/AuditEventKafkaSender.java

```diff
@@ -0,0 +1,186 @@
+/**
+ * Copyright 2019 Pinterest, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.pinterest.singer.loggingaudit.client;
+
+import com.pinterest.singer.loggingaudit.client.common.LoggingAuditClientMetrics;
+import com.pinterest.singer.loggingaudit.client.utils.KafkaUtils;
+import com.pinterest.singer.loggingaudit.thrift.LoggingAuditHeaders;
+import com.pinterest.singer.loggingaudit.thrift.LoggingAuditEvent;
+import com.pinterest.singer.loggingaudit.thrift.LoggingAuditStage;
+import com.pinterest.singer.loggingaudit.thrift.configuration.KafkaSenderConfig;
+import com.pinterest.singer.metrics.OpenTsdbMetricConverter;
+
+import org.apache.kafka.clients.producer.Callback;
+import org.apache.kafka.clients.producer.KafkaProducer;
+import org.apache.kafka.clients.producer.ProducerRecord;
+import org.apache.kafka.clients.producer.RecordMetadata;
+import org.apache.thrift.TException;
+import org.apache.thrift.TSerializer;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.concurrent.ArrayBlockingQueue;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+public class AuditEventKafkaSender implements LoggingAuditEventSender {
+
+  private static final Logger LOG = LoggerFactory.getLogger(AuditEventKafkaSender.class);
+
+  private final LoggingAuditStage stage;
+  private final String host;
+  private ArrayBlockingQueue<LoggingAuditEvent> queue;
+  private KafkaProducer<byte[], byte[]> kafkaProducer;
+  private TSerializer serializer = new TSerializer();
+  private AtomicBoolean cancelled = new AtomicBoolean(false);
+  private String topic;
+  private String name;
+  private Thread thread;
+  private int dequeueWaitInSeconds = 30;
+  private int stopGracePeriodInSeconds = 300;
+  private int callingThreadSleepInSeconds = 30;
+
+
+  public AuditEventKafkaSender(KafkaSenderConfig config,
+                               ArrayBlockingQueue<LoggingAuditEvent> queue,
+                               LoggingAuditStage stage , String host, String name){
+     this.topic = config.getTopic();
+     this.queue = queue;
+     this.stage = stage;
+     this.host = host;
+     this.name = name;
+  }
+
+  public KafkaProducer<byte[], byte[]> getKafkaProducer() {
+    return kafkaProducer;
+  }
+
+  public void setKafkaProducer(KafkaProducer<byte[], byte[]> kafkaProducer) {
+    this.kafkaProducer = kafkaProducer;
+  }
+
+  @Override
+  public void run() {
```

Comment:
As reliable audit message writing is essential to guarantee auditing accuracy, we need to improve writing reliability by having multiple retries and randomly selecting different partitions during retry to minimize the impact of failure on a specific partition.

Reply:
made changes in the update of the PR. also tested with integration test and performance test.